File: src/lib/cpi/signals/hormuz.ts
// signals/hormuz.ts — AIS vessel tracking in Strait of Hormuz
// The alpha signal. Counts physical ships transiting the strait.
// Uses AISStream WebSocket — shortened to 45s for serverless timeout budget.

import type { SignalResult } from "../types";
import {
  AISSTREAM_API_KEY,
  HORMUZ_BOX,
  HORMUZ_LISTEN_SECONDS,
  HORMUZ_NORMAL_VESSELS,
} from "../config";

function noData(reason: string): SignalResult {
  return {
    signal: "hormuz",
    score: 50,
    confidence: 0.0,
    interpretation: `NO DATA: ${reason}`,
    alert: false,
    count: 0,
    flag_count: 0,
    flags: [],
    messages_received: 0,
    timestamp: new Date().toISOString(),
  };
}

function score(count: number, flagCount: number): number {
  if (count === 0) return 5;
  const k = 2.5 / HORMUZ_NORMAL_VESSELS;
  const raw = 100 * (1 - Math.exp(-k * count));
  let s = Math.max(5, Math.min(95, Math.round(raw)));
  if (count >= 3) {
    if (flagCount >= 3) s = Math.min(95, s + 5);
    else if (flagCount === 1) s = Math.max(5, s - 8);
  }
  return s;
}

function interpret(
  count: number,
  flagCount: number,
  s: number
): string {
  const flags = flagCount ? `, ${flagCount} flag states` : "";
  if (count === 0)
    return "Zero vessels in strait \u2014 possible blockade or shipping halt";
  if (s >= 70)
    return `${count} vessels transiting${flags} \u2014 normal commercial flow`;
  if (s >= 40)
    return `${count} vessels transiting${flags} \u2014 reduced traffic`;
  return `${count} vessels transiting${flags} \u2014 severely disrupted`;
}

export async function collectHormuz(): Promise<SignalResult> {
  if (!AISSTREAM_API_KEY) {
    return noData("AISSTREAM_API_KEY not configured");
  }

  const vessels = new Map<
    string,
    { type: number; flag: string }
  >();
  let messagesReceived = 0;
  let connectionOk = false;

  try {
    // Dynamic import — ws may not be available in all environments
    // In Node.js serverless, native WebSocket is available in Node 21+
    // For older runtimes, we use the global WebSocket or ws package
    const WS =
      typeof WebSocket !== "undefined"
        ? WebSocket
        : (await import("ws")).default;

    const ws = new WS("wss://stream.aisstream.io/v0/stream");

    await new Promise<void>((resolve, reject) => {
      const deadline = Date.now() + HORMUZ_LISTEN_SECONDS * 1000;

      ws.onopen = () => {
        connectionOk = true;
        ws.send(
          JSON.stringify({
            APIKey: AISSTREAM_API_KEY,
            BoundingBoxes: [HORMUZ_BOX],
            FilterMessageTypes: ["PositionReport"],
          })
        );
      };

      ws.onmessage = (event: { data: unknown }) => {
        messagesReceived++;
        try {
          const raw =
            typeof event.data === "string"
              ? event.data
              : event.data?.toString?.() ?? "";
          const msg = JSON.parse(raw);
          const meta = msg?.MetaData ?? {};
          const mmsi = meta.MMSI ?? "";
          if (mmsi) {
            vessels.set(String(mmsi), {
              type: meta.ShipType ?? 0,
              flag: meta.country_iso ?? meta.Flag ?? "",
            });
          }
        } catch {
          // Skip malformed messages
        }

        if (Date.now() >= deadline) {
          ws.close();
          resolve();
        }
      };

      ws.onerror = (err: unknown) => {
        const msg =
          err && typeof err === "object" && "message" in err
            ? (err as { message: string }).message
            : String(err);
        reject(new Error(`WebSocket error: ${msg.slice(0, 100)}`));
      };

      ws.onclose = () => resolve();

      // Safety timeout — don't exceed serverless budget
      setTimeout(() => {
        try {
          ws.close();
        } catch {
          /* already closed */
        }
        resolve();
      }, HORMUZ_LISTEN_SECONDS * 1000 + 2000);
    });
  } catch (e) {
    const err = e instanceof Error ? e.message : String(e);
    if (err.includes("403") || err.includes("401")) {
      return noData(`API key rejected \u2014 ${err.slice(0, 80)}`);
    }
    return noData(`WebSocket error \u2014 ${err.slice(0, 80)}`);
  }

  const count = vessels.size;
  const flagSet = new Set<string>();
  for (const v of vessels.values()) {
    if (v.flag) flagSet.add(v.flag);
  }
  const flagCount = flagSet.size;
  const s = score(count, flagCount);
  let interp = interpret(count, flagCount, s);

  let confidence: number;
  if (connectionOk && messagesReceived > 0) {
    confidence = 1.0;
  } else if (connectionOk) {
    confidence = 0.5;
    interp += " (low AIS volume \u2014 coverage gap possible)";
  } else {
    confidence = 0.0;
  }

  return {
    signal: "hormuz",
    score: s,
    confidence,
    interpretation: interp,
    alert: count === 0 && confidence >= 0.5,
    count,
    flag_count: flagCount,
    flags: Array.from(flagSet),
    messages_received: messagesReceived,
    timestamp: new Date().toISOString(),
  };
}

File: cpi/signals/hormuz.py
# signals/hormuz.py — AIS vessel tracking in Strait of Hormuz
# The alpha signal. Counts physical ships transiting the strait.
# More ships = trade flowing = peace. Zero ships = blockade or data failure.
#
# CRITICAL: Distinguishes "no ships" from "can't see."
# confidence=0 if WebSocket fails, confidence=1 if we got real data.

import asyncio
import json
import math
import time
from datetime import datetime, timezone

try:
    import websockets
except ImportError:
    websockets = None

import sys, os
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from config import (
    AISSTREAM_API_KEY, HORMUZ_BOX, HORMUZ_LISTEN_SECONDS,
    HORMUZ_NORMAL_VESSELS
)


def _no_data(reason):
    """Return a result that clearly says 'I have no data.'"""
    return {
        "signal": "hormuz", "score": 50, "confidence": 0.0,
        "interpretation": f"NO DATA: {reason}",
        "alert": False, "count": 0, "flag_count": 0, "flags": [],
        "messages_received": 0,
        "timestamp": datetime.now(timezone.utc).isoformat(),
    }


def _score(count, flag_count):
    """Smooth sigmoid scoring. No step functions, no cliffs."""
    if count == 0:
        return 5
    # Calibrate so HORMUZ_NORMAL_VESSELS → ~75
    k = 2.5 / HORMUZ_NORMAL_VESSELS
    raw = 100 * (1 - math.exp(-k * count))
    s = max(5, min(95, round(raw)))
    # Flag diversity: smooth adjustment, not a cliff
    if count >= 3:
        if flag_count >= 3:
            s = min(95, s + 5)
        elif flag_count == 1:
            s = max(5, s - 8)
    return s


def _interpret(count, flag_count, score):
    flags = f", {flag_count} flag states" if flag_count else ""
    if count == 0:
        return "Zero vessels in strait — possible blockade or shipping halt"
    if score >= 70:
        return f"{count} vessels transiting{flags} — normal commercial flow"
    if score >= 40:
        return f"{count} vessels transiting{flags} — reduced traffic"
    return f"{count} vessels transiting{flags} — severely disrupted"


async def collect_hormuz(duration_seconds=None):
    if duration_seconds is None:
        duration_seconds = HORMUZ_LISTEN_SECONDS

    if not AISSTREAM_API_KEY:
        return _no_data("AISSTREAM_API_KEY not configured")
    if websockets is None:
        return _no_data("websockets library not installed (pip install websockets)")

    vessels = {}
    messages_received = 0
    connection_ok = False

    try:
        subscribe = {
            "APIKey": AISSTREAM_API_KEY,
            "BoundingBoxes": [HORMUZ_BOX],
            "FilterMessageTypes": ["PositionReport"],  # Only position data, skip static reports
        }
        async with websockets.connect(
            "wss://stream.aisstream.io/v0/stream",
            ping_interval=20, ping_timeout=10
        ) as ws:
            await ws.send(json.dumps(subscribe))
            connection_ok = True
            deadline = time.time() + duration_seconds

            while time.time() < deadline:
                try:
                    raw = await asyncio.wait_for(ws.recv(), timeout=5)
                    messages_received += 1
                    msg = json.loads(raw)
                    meta = msg.get("MetaData", {})
                    mmsi = meta.get("MMSI", "")
                    if not mmsi:
                        continue
                    vessels[mmsi] = {
                        "type": meta.get("ShipType", 0),
                        "flag": meta.get("country_iso", meta.get("Flag", "")),
                    }
                except asyncio.TimeoutError:
                    continue
                except websockets.exceptions.ConnectionClosed:
                    print(f"[HORMUZ] WebSocket closed after {messages_received} msgs")
                    break

    except Exception as e:
        err = str(e)[:100]
        if "403" in err or "401" in err:
            return _no_data(f"API key rejected — {err}")
        if "proxy" in err.lower():
            return _no_data(f"proxy blocked — {err}")
        return _no_data(f"WebSocket error — {err}")

    count = len(vessels)
    flags = list(set(v["flag"] for v in vessels.values() if v["flag"]))
    flag_count = len(flags)
    score = _score(count, flag_count)
    interp = _interpret(count, flag_count, score)

    # Confidence logic:
    # 1.0 = connected, received messages, have real vessel data
    # 0.5 = connected, zero messages (AIS coverage gap?)
    # 0.0 = couldn't connect (handled above)
    if connection_ok and messages_received > 0:
        confidence = 1.0
    elif connection_ok:
        confidence = 0.5
        interp += " (low AIS volume — coverage gap possible)"
    else:
        confidence = 0.0

    return {
        "signal": "hormuz", "score": score, "confidence": confidence,
        "interpretation": interp,
        "alert": count == 0 and confidence >= 0.5,
        "count": count, "flag_count": flag_count, "flags": flags,
        "messages_received": messages_received,
        "timestamp": datetime.now(timezone.utc).isoformat(),
    }
